File: csharp/Exo4/Game.cs
//========================================================================
// This conversion was produced by the Free Edition of
// Java to C# Converter courtesy of Tangible Software Solutions.
// Order the Premium Edition at https://www.tangiblesoftwaresolutions.com
//========================================================================

using System.Text;

namespace Soat.CleanCoders.Refactoring.Exo4
{
	public class Game
	{

		public StringBuilder board;

		public Game(string s)
		{
			board = new StringBuilder(s);
		}

		private Game(StringBuilder s, int position, char player)
		{
			board = new StringBuilder(s);
			board.Append(s);
			board[position] = player;
		}

		public virtual int move(char player)
		{
			for (int i = 0; i < 9; i++)
			{
				if (board[i] == '-')
				{
					Game game = play(i, player);
					if (game.winner() == player)
					{
						return i;
					}
				}
			}

			for (int i = 0; i < 9; i++)
			{
				if (board[i] == '-')
				{
					return i;
				}
			}
			return -1;
		}

		public virtual char winner()
		{
			if (board[0] != '-' && board[0] == board[1] && board[1] == board[2])
			{
				return board[0];
			}

			if (board[3] != '-' && board[3] == board[4] && board[4] == board[5])
			{
				return board[3];
			}

			if (board[6] != '-' && board[6] == board[7] && board[7] == board[8])
			{
				return board[6];
			}

			return '-';
		}

		private Game play(int i, char player)
		{
			return new Game(this.board, i, player);
		}
	}

}

File: javascript/app/exo4/game.js
function Game(s) {
	this.move = move;
	this.winner = winner;
	
	let board = s;
	let initBoard = s;
	
	function Game(s, position, player) {
		initBoard = s;
		board = s;
		board = board.concat(s);
		board = board.replaceAt(position, player);
	}
	
	function move(player) {
		for (let i = 0; i < 9; i++) {
			if (board.charAt(i) === '-') {
				let game = play(i, player);
				if (game.winner() === player)
					return i;
			}
		}
		
		for (let i = 0; i < 9; i++) {
			if (initBoard.charAt(i) == '-')
				return i;
		}
		return -1;
	}
	
	function winner() {
		if (board.charAt(0) != '-' && board.charAt(0) == board.charAt(1) && board.charAt(1) == board.charAt(2))
			return board.charAt(0);
		
		if (board.charAt(3) != '-' && board.charAt(3) == board.charAt(4)  && board.charAt(4) == board.charAt(5))
			return board.charAt(3);
		
		if (board.charAt(6) != '-' && board.charAt(6) == board.charAt(7)   && board.charAt(7) == board.charAt(8))
			return board.charAt(6);
		
		return '-';
	}
	
	
	function play(i, player) {
		let game = new Game(initBoard, i, player);
		return game;
	}
	
	Game.prototype.winner = winner;
	
	String.prototype.replaceAt = function(index, char) {
		let a = this.split("");
		a[index] = char;
		return a.join("");
	}
}

module.exports = Game;